File: app/src/main/java/com/udacity/sandwichclub/ui/SandwichFragment.java
package com.udacity.sandwichclub.ui;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.udacity.sandwichclub.R;
import com.udacity.sandwichclub.model.Sandwich;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SandwichFragment extends Fragment {
    /* Class Constants */
    private static final String ARG_SANDWICH = "sandwich";
    private static final String TAG = SandwichFragment.class.getSimpleName();

    /* Class variables */
    @BindView(R.id.sandwich_image_view)
    ImageView sandwichImageView;
    @BindView(R.id.name_text_view)
    TextView nameTextView;
    @BindView(R.id.aka_text_view)
    TextView akaTextView;
    @BindView(R.id.origin_place_text_view)
    TextView originTextView;
    @BindView(R.id.desc_text_view)
    TextView descTextView;
    @BindView(R.id.ingredients_text_view)
    TextView ingredientsTextView;
    private Sandwich sandwich;

    public static SandwichFragment newInstance(Sandwich sandwich) {
        Bundle args = new Bundle();
        args.putParcelable(ARG_SANDWICH, sandwich);

        SandwichFragment fragment = new SandwichFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getArguments() != null) {
            sandwich = getArguments().getParcelable(ARG_SANDWICH);
        } else {
            //noinspection ConstantConditions
            Log.d(TAG, "Sandwich not received from " + getActivity().getLocalClassName());
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_detail, container, false);
        ButterKnife.bind(this, view);

        populateUI();
        RequestOptions options = new RequestOptions()
                .centerCrop()
                .error(R.drawable.ic_error);

        //noinspection ConstantConditions
        Glide.with(getActivity())
                .setDefaultRequestOptions(options)
                .load(sandwich.getImage())
                .into(sandwichImageView);

        return view;
    }

    private void populateUI() {
        nameTextView.setText(parseString(sandwich.getMainName()));
        akaTextView.setText(parseString(sandwich.getAlsoKnownAs()));
        originTextView.setText(parseString(sandwich.getPlaceOfOrigin()));
        descTextView.setText(parseString(sandwich.getDescription()));
        ingredientsTextView.setText(parseString(sandwich.getIngredients()));
    }

    private String parseString(List<String> listOfWords) {
        String listOfOtherNames = null;
        if (listOfWords.size() > 0) {
            listOfOtherNames = TextUtils.join(", ", listOfWords);
        } else {
            return getString(R.string.detail_unavailable_error_message);
        }

        return listOfOtherNames;
    }

    private String parseString(String string) {
        if (TextUtils.isEmpty(string)) {
            return getString(R.string.detail_unavailable_error_message);
        }

        return string;
    }
}

File: app/src/main/java/com/udacity/sandwichclub/ui/SandwichFragment.kt
package com.udacity.sandwichclub.ui

import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.udacity.sandwichclub.R
import com.udacity.sandwichclub.databinding.FragmentSandwichBinding
import com.udacity.sandwichclub.model.Sandwich


class SandwichFragment : Fragment() {

    companion object {
        const val SANDWICH = "sandwich"
    }

    private var _binding: FragmentSandwichBinding? = null

    private lateinit var sandwichImageView: ImageView
    private lateinit var nameTextView: TextView
    private lateinit var akaTextView: TextView
    private lateinit var originTextView: TextView
    private lateinit var descTextView: TextView
    private lateinit var ingredientsTextView: TextView
    private lateinit var sandwich: Sandwich

    private val binding get() = _binding!!
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        arguments?. let {
            sandwich = it.getParcelable(SANDWICH)!!
        }
    }

    @Suppress("RedundantNullableReturnType")
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        _binding = FragmentSandwichBinding.inflate(inflater, container, false)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        sandwichImageView = binding.sandwichImageView
        nameTextView = binding.nameTextView
        akaTextView = binding.akaTextView
        originTextView = binding.originPlaceTextView
        descTextView = binding.descTextView
        ingredientsTextView = binding.ingredientsTextView

        populateUI()
        val options = RequestOptions().centerCrop().error(R.drawable.ic_error)

        Glide.with(view.context)
            .setDefaultRequestOptions(options)
            .load(sandwich.image)
            .into(sandwichImageView)
    }

    private fun populateUI() {
        nameTextView.text = parseString(sandwich.mainName)
        akaTextView.text = parseString(sandwich.alsoKnownAs)
        originTextView.text = parseString(sandwich.placeOfOrigin)
        descTextView.text = parseString(sandwich.description)
        ingredientsTextView.text = parseString(sandwich.ingredients)
    }

    private fun parseString(listOfWords: List<String>): String {
        var listOfOtherNames = ""
        if (listOfWords.isNotEmpty()) {
            listOfOtherNames = TextUtils.join(", ", listOfWords);
        } else {
            return getString(R.string.detail_unavailable_error_message);
        }

        return listOfOtherNames
    }

    private fun parseString(string: String): String {
        if (TextUtils.isEmpty(string)) {
            return getString(R.string.detail_unavailable_error_message)
        }

        return string
    }
}